federation/pkg/federation-controller/util/federated_informer.go

```diff
@@ -0,0 +1,408 @@
+	// are synced. This is only a basic check whether the data inside of the store is usable.
+	// It is not a full synchronization/locking mechanism it only tries to ensure that out-of-sync
+	// issues occur less often.	All users of the interface should assume
+	// that there may be significant delays in content updates of all kinds and write their
+	// code that it doesn't break if something is slightly out-of-sync.
+	Synced(clusters []string) bool
+}
+
+// An interface to access federation members and clients.
+type FederationView interface {
+	// GetClientsetForCluster returns a clientset for the cluster, if present.
+	GetClientsetForCluster(clusterName string) (federation_release_1_4.Interface, error)
+
+	// GetClusers returns all clusters for which the sub-informers are run.
+	GetClusters() []*federation_api.Cluster
+
+	// GetReadyCluster returns the cluster with the given name, if found.
+	GetCluster(name string) (*federation_api.Cluster, bool)
+
+	// Synced returns true if the view is synced (for the first time)
+	ClusterSynced() bool
+}
+
+// A tyoe of structure that combines an infomer running agains federated api server and listening for cluster updates
+// with multiple Kubernetes API informers (called target infomers) running against federation members. Whenever a new
+// cluster is added to the federation an infomer is created for it using TargetInformerFactory. Infomers are stoped
+// when a cluster is either put offline of deleted. It is assumed that some controller keeps an eye on the cluster list
+// and thus the clusters in ETCD are up to date.
+type FederatedInfomer interface {
+	FederationView
+
+	// Returns a store created over all stores from target informers.
+	GetTargetStore() FederatedReadOnlyStore
+
+	// Starts all the processes.
+	Start()
+
+	// Stops all the processes inside the infomer.
+	Stop()
+}
+
+// The function that should be used to create an infomer on the target object. Store should use
+// framework.DeletionHandlingMetaNamespaceKeyFunc as a keying function.
+type TargetInformerFactory func(*federation_api.Cluster, federation_release_1_4.Interface) (cache.Store, *framework.Controller)
+
+// Builds a FederatedInformer for the given federation client and factory.
+func NewFederatedInformer(federationClient federation_release_1_4.Interface, targetInformerFactory TargetInformerFactory) FederatedInfomer {
+
+	federatedInformer := &federatedInformerImpl{
+		targetInformerFactory: targetInformerFactory,
+		clientFactory: func(cluster *federation_api.Cluster) (federation_release_1_4.Interface, error) {
+			clusterConfig, err := BuildClusterConfig(cluster)
+			if clusterConfig != nil {
+				clientset := federation_release_1_4.NewForConfigOrDie(restclient.AddUserAgent(clusterConfig, UserAgentName))
+				return clientset, nil
+			}
+			return nil, err
+		},
+		targetInformers: make(map[string]informer),
+	}
+
+	isClusterReady := func(cluster *federation_api.Cluster) bool {
```

Comment:
Please make this a named function (this can potentially be useful not only inside this method).